traits book update (#843)
openmls/openmls: book/src/traits/traits.md

# OpenMLS Traits

> **⚠️  These traits are responsible for all cryptographic operations and randomness
> within OpenMLS.
> Please ensure you know what you're doing when implementing your own versions.**

Because implementing the `OpenMLSCryptoProvider` is challenging, requires

> tremendous care, and the average OpenMLS consumer does not want (or should) do,

Reviewer: I'm having trouble parsing that sentence. Maybe the following?
```suggestion
tremendous care, and is not what the average OpenMLS consumer wants to (or should) do,
```

> we provide two implementations that can be used.

Reviewer 2: aka "Do not roll your own!" 🙂

- [Rust Crypto]
- [Evercrypt]

**Rust Crypto Provider**

> The go-to default at the moment is an implementation using common, native Rust

Reviewer: ```suggestion
The go-to default at the moment is an implementation using commonly used, native Rust
```

### OpenMlsRand

This trait defines two functions to generate arrays and vectors, and is used by
OpenMLS to generate randomness for key generation and random identifiers.

> While there is the commonly used [rand crate] not all implementation use it.

Reviewer: ```suggestion
While there is the commonly used [rand crate] not all implementations use it.
```

> OpenMLS therefore defines it's own randomness trait that needs to be implemented

Reviewer: ```suggestion
OpenMLS therefore defines its own randomness trait that needs to be implemented
```

{{#include ../../../traits/src/key_store.rs:15:40}}
```

> **NOTE:** Right now key material need to be extractable from the key store.

Reviewer: ```suggestion
**NOTE:** Right now key material needs to be extractable from the key store.
```

> This will most likely change in future.

Reviewer 2: ```suggestion
This will most likely change in the future.
```

### OpenMlsCryptoProvider

Additionally, there's a wrapper trait defined that is expected to be passed into
the public OpenMLS API.
Some OpenMLS APIs require only one of the sub-traits though.

```rust,no_run,noplayground
{{#include ../../../traits/src/traits.rs:15:28}}
```

## Implementation Notes

It is not necessary to implement all sub-traits if one functionality is missing.

> If you want to use a persisting key store for example, it is sufficient to do a new implementation of the key store trait and combine it with one of the provided crypto and randomness trait implementations.

Reviewer: Would that be a reason to make the `memory_keystore` public again?

Author: 🤔  it is a published crate. So how do you think it is not public?

Reviewer: Never mind. I got two things confused, I think.
